fx.go

```diff
@@ -19,3 +19,149 @@
 // THE SOFTWARE.
 
 package fx // import "go.uber.org/fx"
+
+import (
+	"io"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/uber-go/tally"
+
+	"go.uber.org/dig"
+	"go.uber.org/fx/config"
+	"go.uber.org/fx/metrics"
+	"go.uber.org/fx/ulog"
+	"go.uber.org/zap"
+)
+
+// Component is a function that initializes and returns objects to be shared across the code
+type Component interface{}
+
+// Module is a building block or primary function performed by a service built on fx
+type Module interface {
+	Name() string
+	Constructor() []Component
```

Comment:
That's not true. Taking configuration as part of the component means that a module can not control the *number* of modules provided based on the configuration.

For example, consider that yarpc module would automatically expose all the clients in the graph:

```
modules:
  yarpc:
    outbounds:
    - foo
    - bar
```

We'd want to expose `foo` and `bar` in the graph based on the configuration provided to the initial module and dynamically expose two components.

Otherwise, how would a module react to the configuration and dynamically expose various numbers of components?